Feature/s120/bee 7332/route optimization by distance

iFHA opened the pull request:
Foram adicionadas duas formas de otimizar waypoins por distância:
1 - Matriz de distância(Route Matrix API)
2 - Route Optimization API

Commit:
feat(BEE-7332): waypoints optimization by distance from route matrix

Commit:
feat(BEE-7332): implemented route optimization api and waypoints optimizer by travel time methods

Commit:
fix(BEE-7332): add duration field mask to route matrix request

Commit:
fix(BEE-7332): logic error

Commit:
fix(BEE-7332): syntax error

Commit:
fix(BEE-7332): correct route optimization by distance using route matrix api

Commit:
fix(BEE-7332): correct route optimization api config name

Commit:
fix(BEE-7332): correct service account credentials config path

Commit:
fix(BEE-7332): correct project_id config path

Commit:
fix(BEE-7332): correct project_id config path

Commit:
feat(BEE-7332): add optimization type in return

Commit:
feat(BEE-7332): correct route optimization type return

Commit:
chore(BEE-7332): ajust some methods summaries

chrystianpaulino commented on Utils/HelpersRouteMatrix.php:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    private function validateRequest(array $origins, array $destinations): void
+    {
```
Atualmente está usando Exception. Seria mais semântico definir uma exception específica:

`use InvalidArgumentException;
...
throw new InvalidArgumentException('You must provide at least one origin');
`

chrystianpaulino commented on Utils/HelpersRouteMatrix.php:
```diff
@@ -0,0 +1,84 @@
+     * @param RouteTravelModeEnum $mode
+     * @return array{origins: array, destinations: array, travelMode: string}
+     */
+    public function formatRequest(array $origins, array $destinations, RouteTravelModeEnum $mode = RouteTravelModeEnum::DRIVE): array
```
formatRequest() retorna um array com estrutura conhecida. Poderia retornar um DTO próprio (RouteMatrixRequestDTO), o que garantiria mais segurança e legibilidade.

chrystianpaulino commented on Utils/HelpersRouteOptimization.php:
```diff
@@ -0,0 +1,111 @@
+        $url = config('googlemaps.route_optimization_api.url');
+        $projectId = config('googlemaps.route_optimization_api.service_account_credentials.project_id');
+        
+        if(empty($url)) {
```
Uso de exceções genéricas
Sugestão 1: Substituir Exception por algo mais semântico, como InvalidArgumentException ou uma exception custom (RouteOptimizationException).
Sugestão 2: Evite passar o código HTTP diretamente no construtor da Exception — use um atributo próprio se necessário.

chrystianpaulino commented on Utils/HelpersRouteOptimization.php:
```diff
@@ -0,0 +1,111 @@
+            "searchMode" => "SEARCH_MODE_UNSPECIFIED",
+            "considerRoadTraffic" => false
+        ];
+
```
O trait usa $this->type, mas ela não é garantida.

`if (!isset($this->type)) {
    throw new LogicException('Route optimization type must be defined before calling formatRequest()');
}
`

chrystianpaulino commented on Utils/HelpersRouteOptimization.php:
```diff
@@ -0,0 +1,111 @@
+     * @return OptimizedWaypointsDTO
+     */
+    public function formatResponse(array $response): OptimizedWaypointsDTO
+    {
```
Acessar índices diretamente ($response['metrics']['aggregatedRouteMetrics']) pode gerar warnings se a estrutura mudar. Use data_get() (do Laravel) ou Arr::get():

`$aggregated = data_get($response, 'metrics.aggregatedRouteMetrics', []);
`

chrystianpaulino commented on src/Factories/WaypointsOptimizerFactory.php:
```diff
@@ -0,0 +1,24 @@
+                config('googlemaps.waypoints_optimizer_by_distance_preferred_api') === 'route_matrix' ?
+                new WaypointsOptimizerByRouteMatrixDistance():
+                new WaypointsOptimizerByRouteOptimizationApi(),
+            default => throw new Exception('Invalid waypoints optimizer type, accepted values are: MIN_TRAVEL_TIME, MIN_DISTANCE'),
```
use InvalidArgumentException;
...
default => throw new InvalidArgumentException(
    'Invalid waypoints optimizer type. Accepted: MIN_TRAVEL_TIME, MIN_DISTANCE.'
),

chrystianpaulino commented on src/Services/Impl/WaypointsOptimizerByRouteMatrixDistance.php:
```diff
@@ -0,0 +1,136 @@
+        $waypointsCount = count($origins);
+        $destinationWaypointIndex = $waypointsCount - 1;
+        if ($waypointsCount < 4) {
+            throw new Exception('You must provide at least two intermediate waypoints');
```
Substituir Exception genérica
Use algo mais expressivo, como InvalidArgumentException ou crie uma específica (RouteMatrixOptimizationException), especialmente porque essas exceções são de domínio, não erros genéricos de execução.

chrystianpaulino commented on src/Services/Impl/WaypointsOptimizerByRouteMatrixDistance.php:
```diff
@@ -0,0 +1,136 @@
+{
+    private RouteMatrix $routeMatrix;
+    private RouteTravelModeEnum $mode;
+    public function __construct()
```
Prefira injeção de dependência para permitir testes e mock:

`public function __construct(RouteMatrix $routeMatrix = null)
{
    $this->routeMatrix = $routeMatrix ?? new RouteMatrix();
    $this->mode = RouteTravelModeEnum::DRIVE;
}
`

chrystianpaulino commented on src/Services/Impl/WaypointsOptimizerByRouteMatrixDistance.php:
```diff
@@ -0,0 +1,136 @@
+        $origins = $destinations = [$optimizeWaypointsDTO->origin, ...$optimizeWaypointsDTO->intermediateWaypoints, $optimizeWaypointsDTO->destination];
+        $waypointsCount = count($origins);
+        $destinationWaypointIndex = $waypointsCount - 1;
+        if ($waypointsCount < 4) {
```
As mensagens de erro podem ser mais informativas:

`if ($waypointsCount < 4) {
    throw new InvalidArgumentException("At least two intermediate waypoints are required (got {$waypointsCount}).");
}
`

chrystianpaulino commented on src/Services/Impl/WaypointsOptimizerByRouteMatrixDistance.php:
```diff
@@ -0,0 +1,136 @@
+            $optimizedWaypointsOrdered[] = $index;
+        }
+
+        $dataFromBetweenLastIntermediateWaypointAndDestination = $this->getDataFromRouteMatrixBetweenTwoWaypoints($index, $destinationWaypointIndex, $routeMatrix);
```
poderia ter um nome mais semântico como getDataFromRouteMatrixBetweenTwoWaypoints

chrystianpaulino commented on src/Services/Impl/WaypointsOptimizerByRouteMatrixDistance.php:
```diff
@@ -0,0 +1,136 @@
+        );
+    }
+
+    private function getClosestWaypointTo(int $originWaypointIndex, array $routeMatrix, int $destinationWaypointIndex, array $visitedIndexes): RouteMatrixResponseDTO
```
O método getClosestWaypointTo() percorre todo o routeMatrix várias vezes. Se a matriz for grande (origens × destinos), isso impacta desempenho. Uma melhoria possível:
Pré-agrupar o routeMatrix por originIndex em cache (ex: $this->routesByOrigin[$originIndex]), reduzindo buscas repetidas ou converter routeMatrix em um mapa: $matrixMap[$originIndex][$destinationIndex] = $item;

chrystianpaulino commented on src/Services/Impl/WaypointsOptimizerByRouteMatrixDistance.php:
```diff
@@ -0,0 +1,136 @@
+
+    private function getClosestWaypointTo(int $originWaypointIndex, array $routeMatrix, int $destinationWaypointIndex, array $visitedIndexes): RouteMatrixResponseDTO
+    {
+        $distancesFromOrigin = array_filter($routeMatrix, function (RouteMatrixResponseDTO $item) use ($originWaypointIndex, $visitedIndexes, $destinationWaypointIndex) {
```
Se array_filter não encontrar nada, o usort roda sobre array vazio e depois acessa [0] — risco de Undefined offset.

`if (empty($distancesFromOrigin)) {
    throw new Exception("No valid routes found from waypoint {$originWaypointIndex}");
}
`

chrystianpaulino commented on src/Services/Impl/WaypointsOptimizerByRouteMatrixDistance.php:
```diff
@@ -0,0 +1,136 @@
+     * @return OptimizedWaypointsDTO
+     * @throws Exception when the route matrix response is null or the response code is not 200
+     */
+    private function getOptimizedWaypointsByDistanceFromRouteMatrix(array $routeMatrix, int $destinationWaypointIndex, int $intermediateWaypointsCount): OptimizedWaypointsDTO
```
getOptimizedWaypointsByDistanceFromRouteMatrix() faz tudo: resolve caminho, soma métricas e cria DTO.

Poderia ser dividido em métodos menores:
calculateOptimalOrder()
calculateTotalMetrics()
buildOptimizedDto()

chrystianpaulino commented on src/Services/Impl/WaypointsOptimizerByRouteMatrixDistance.php:
```diff
@@ -0,0 +1,136 @@
+
+    private function adjustOptimizedWaypointsIndexes(array $optimizedWaypointsOrdered): array
+    {
+        return array_map(fn (int $index) => $index - 1, $optimizedWaypointsOrdered);
```
O método adjustOptimizedWaypointsIndexes() reduz todos os índices em 1, inclusive se algum índice já for 0.
Isso pode gerar -1. Sugiro adicionar uma verificação:

return array_map(fn(int $index) => max(0, $index - 1), $optimizedWaypointsOrdered);

chrystianpaulino commented on src/Services/Impl/WaypointsOptimizerByRouteOptimizationApi.php:
```diff
@@ -0,0 +1,47 @@
+        ->fetchAccessTokenWithAssertion()['access_token'] ?? '';
+        
+        if (empty($token)) {
+            throw new Exception('Failed to get access token for route optimization');
```
exception....

Commit:
feat(BEE-7332): add distance-based route optimization using TSP algorithm

- Implement RouteMatrixDistanceOptimizerByTSP with TSP optimization
- Add OptimizedWaypointsSummaryDTO and related exceptions
- Update route matrix helpers with validation
